pkg/daemon/daemon.go

```diff
@@ -1325,44 +1385,83 @@ func (dn *Daemon) triggerUpdateWithMachineConfig(currentConfig, desiredConfig *m
-	}
-	// And the rest of the disk state
-	// We want to verify the disk state in the spec version that it was created with,
-	// to remove possibilities of behaviour changes due to translation
-	ignconfigi, err := ctrlcommon.IgnParseWrapper(currentConfig.Spec.Config.Raw)
-	if err != nil {
-		return errors.Errorf("Failed to parse Ignition for validation: %s", err)
-	}
+	// Make sure only a single instance of this can run at a time
+	dn.diskValidatorLock.Lock()
+	defer dn.diskValidatorLock.Unlock()
 
-	switch typedConfig := ignconfigi.(type) {
-	case ign3types.Config:
-		if err := checkV3Files(ignconfigi.(ign3types.Config).Storage.Files); err != nil {
-			return err
-		}
-		if err := checkV3Units(ignconfigi.(ign3types.Config).Systemd.Units); err != nil {
-			return err
+	validateFunc := func() error {
+		// Be sure we're booted into the OS we expect
+		osMatch := dn.checkOS(currentConfig.Spec.OSImageURL)
+		if !osMatch {
+			return errors.Errorf("expected target osImageURL %q, have %q", currentConfig.Spec.OSImageURL, dn.bootedOSImageURL)
 		}
-		return nil
-	case ign2types.Config:
-		if err := checkV2Files(ignconfigi.(ign2types.Config).Storage.Files); err != nil {
-			return err
+		// And the rest of the disk state
+		// We want to verify the disk state in the spec version that it was created with,
+		// to remove possibilities of behaviour changes due to translation
+		ignconfigi, err := ctrlcommon.IgnParseWrapper(currentConfig.Spec.Config.Raw)
+		if err != nil {
+			return errors.Errorf("Failed to parse Ignition for validation: %s", err)
 		}
-		if err := checkV2Units(ignconfigi.(ign2types.Config).Systemd.Units); err != nil {
-			return err
+
+		switch typedConfig := ignconfigi.(type) {
+		case ign3types.Config:
+			if err := checkV3Files(ignconfigi.(ign3types.Config).Storage.Files); err != nil {
+				return err
+			}
+			if err := checkV3Units(ignconfigi.(ign3types.Config).Systemd.Units); err != nil {
+				return err
+			}
+			return nil
+		case ign2types.Config:
+			if err := checkV2Files(ignconfigi.(ign2types.Config).Storage.Files); err != nil {
+				return err
+			}
+			if err := checkV2Units(ignconfigi.(ign2types.Config).Systemd.Units); err != nil {
+				return err
+			}
+			return nil
+		default:
+			return errors.Errorf("unexpected type for ignition config: %v", typedConfig)
 		}
-		return nil
-	default:
-		return errors.Errorf("unexpected type for ignition config: %v", typedConfig)
 	}
+
+	start := time.Now()
+
+	if err := validateFunc(); err != nil {
+		wErr := fmt.Errorf("on-disk state mismatch: %v", err)
+		dn.recorder.Eventf(getNodeRef(dn.node), corev1.EventTypeWarning, "OnDiskStateValidationFailed", wErr.Error())
+		return wErr
+	}
+
+	end := time.Since(start)
+
+	glog.Infof("Validated on-disk state (took %s)", end)
+	dn.recorder.Eventf(getNodeRef(dn.node), corev1.EventTypeNormal, "OnDiskStateValidationSuccess", "On-disk configuration state validated to equal %s (took %s).", currentConfig.GetName(), end)
```

Comment:
For general eventing: How often is this check running, in a long lived cluster I don't think we want to fill event log with success if the check is happening often.

On failure yes, but on success? Maybe not.